gossip/infostore.go

```diff
@@ -146,26 +152,29 @@ func (is *infoStore) getInfo(key string) *info {
 // addInfo adds or updates an info in the infos map.
 //
 // Returns nil if info was added; error otherwise.
-func (is *infoStore) addInfo(key string, i *info) error {
+func (is *infoStore) addInfo(key string, i *Info) error {
 	// Only replace an existing info if new timestamp is greater, or if
 	// timestamps are equal, but new hops is smaller.
 	var contentsChanged bool
 	if existingInfo, ok := is.Infos[key]; ok {
-		if i.Timestamp < existingInfo.Timestamp ||
-			(i.Timestamp == existingInfo.Timestamp && i.Hops >= existingInfo.Hops) {
+		if iNanos, existingNanos := i.Value.Timestamp.WallTime, existingInfo.Value.Timestamp.WallTime; iNanos < existingNanos ||
+			(iNanos == existingNanos && i.Hops >= existingInfo.Hops) {
 			return util.Errorf("info %+v older than current info %+v", i, existingInfo)
 		}
-		contentsChanged = !reflect.DeepEqual(existingInfo.Val, i.Val)
+		contentsChanged = !gogoproto.Equal(&i.Value, &existingInfo.Value)
```

Comment:
You might as well use the checksum here too, comparing the checksum first will give you an fairly good test to see when if values are different.  Sadly, if it is the same, you'll still need to do full compare the values themselves, but it should save some processing time.